### PR TITLE
DEPS: Add warning if pyarrow is not installed

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

~~Not sure how to add a test for this - importing our warning asserters would cause the warning the be raised! 🤣 ~~

EDIT: Nvm, figured it out.

(I've checked by hand that the warning is both raised when you import pandas directly and import a submodule of pandas e.g. ``pandas.arrays``.
I also double checked that the warning only occurs once and not on every import.)

### CODE REVIEW

File: pandas/__init__.py
Comment (datapythonista) [inline pandas/__init__.py:206]: Is there a reason to not run this this at the end with the other modules? Seems like it's easier to maintain if there is all the clean up at the end or next to the import, but no big deal

File: pandas/__init__.py
Comment (datapythonista) [inline pandas/__init__.py:212]: Isn't warnings already imported at the very top of the file?

File: pandas/__init__.py
Comment (datapythonista) [inline pandas/__init__.py:213]: This will make `pandas.VERSIONS` available when PyArrow is not installed. Do we want to do a `del VERSIONS` inside the `if`?

File: pandas/__init__.py
Comment (datapythonista) [inline pandas/__init__.py:217]: It adds a bit of extra complexity, but I'd personally add another `if` and show different warnings when PyArrow is not installed and when it's too old. I think the displayed warning will be much more clear and concise. Just a personal opinion, no big deal if you prefer to keep it this way.

**Review by datapythonista — COMMENTED** [review]
Looks good, added some minor comments/suggestions.

File: pandas/__init__.py
Comment (mroeschke) [inline pandas/__init__.py:217]:
```suggestion
Pyarrow will become a future required dependency of pandas in the next major release,
```

I would also add a short reason why it's becoming a required dependency

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py:213]: Wow - didn't know that happened.

Thanks for the catch.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: Added a short description talking about more performant data types and better interop.

Comment (lithomas1) [reply pandas/__init__.py]: Updated.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: Thanks, updated.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py:206]: No, not really. 

Personally I like to keep the del closed to where it's being used, but no strong preference.

Comment (datapythonista) [reply pandas/__init__.py:206]: I think `os` is not being used in the code before beibg closed. As said, not important, but to me personally it makes sense to open and close around the block where it is being used, or open and close at the beginning and the end if it's an import that is used in several places. Now `os` is being opened at the beginning, and closed in the middle, in a kind of random place if I'm not wrong. No huge difference, but I think it'll help maintain the code if closed at the end (or we move the import and del around where it is being used).

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: Thx

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: I put the os delete with the deletion of warnings/pa_version_under10p1.

**Comment by lithomas1** [comment]
also cc @pandas-dev/pandas-core for feedback on the warning message itself.

File: pandas/__init__.py
Comment (jorisvandenbossche) [inline pandas/__init__.py:226]:
```suggestion
Pyarrow will become a required dependency of pandas in the next major release of pandas (3.0),
```

I think we can be explicit about what this next major release will be

File: pandas/__init__.py
Comment (jorisvandenbossche) [inline pandas/__init__.py:233]: Do we want to make this stacklevel=1? 

Because this code is top-level and not inside a function or method, it one less than the typical stacklevel of 2 we use in a direct function/method. 
Setting it as 1 would avoid that it triggers a warning when you call a function that has an inline `import pandas` (not sure how important this is, though)

File: pandas/__init__.py
Comment (jorisvandenbossche) [inline pandas/__init__.py:227]: I would maybe explicitly mention the new default string dtype, since that's still the only one that is used by default that requires pyarrow

File: pandas/__init__.py
Comment (Dr-Irv) [inline pandas/__init__.py:209]: IMHO, the code in `pandas/compat/pyarrow.py` should set a value of `pa_not_found` (or something like that) and that should be used in the test.  When I read this code, I didn't understand why you were testing a version number of pyarrow, so I think adding an additional value would make it clear that we are testing for the presence of pyarrow here.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py:226]: > Is the next major release also the next minor release, i.e., 2.2 then 3.0? I noticed that the dev wheels recently switched to reporting `3.0.0.dev0` recently.

No, 2.2 is coming out sometime this week.
But yes, 3.0.0 will be the release after that.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py:230]: The warning doesn't pop up in the REPL when I import pandas if I do stacklevel=1.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: Updated.

File: pandas/__init__.py
Comment (lithomas1) [reply pandas/__init__.py]: Thx, simplified.

**Review by Dr-Irv — APPROVED** [review]
I'm good with it.  Will let others make the final decision.

File: pandas/__init__.py
Comment (jorisvandenbossche) [reply pandas/__init__.py:230]: Hmm, OK forget this. I did test it, but should have done something wrong because now I can't reproduce getting the warning with plain import and stacklevel 1

**Review by datapythonista — APPROVED** [review]
Very nice @lithomas1, looks perfect

File: pandas/tests/test_common.py
Comment (mroeschke) [inline pandas/tests/test_common.py:270]: Could we import this at the top?

**Review by mroeschke — COMMENTED** [review]
Looks like `ci/deps/actions-311-numpydev.yaml` looks like it needs pyarrow otherwise LGTM

**Comment by lithomas1** [comment]
Merging for now, happy to address any followups in another PR.

**Comment by ml-evs** [comment]
Sorry to comment here rather than an issue but this feels quite niche -- just a heads up that the warning added here starts with a newline which makes it annoying to filter out (as `filterwarnings` always matches the start of the message).

e.g., neither

```
"ignore:Pyarrow will become a required dependency of pandas.*:DeprecationWarning"
```

nor

```
"ignore:.*Pyarrow will become a required dependency of pandas.*:DeprecationWarning"
```

nor

```
"ignore:.*pandas.*:DeprecationWarning"
```

will match the message, and what is required is

```
"ignore:\\nPyarrow will become a required dependency of pandas.*:DeprecationWarning"
```

I guess changing it now might cause more pain, but I suspect you may get a few more questions on this before v3 (unless I am singularly confused by this -- a quick GH search shows a few PRs that also took a while to get this correct in other packages)!

**Pull request merged**
